### PR TITLE
ns for fx: add utils for l2 error and cosine similarity

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61380**
* #61377
* #61360
* #61359
* #61358
* #61357
* #61323
* #61129

Summary:

Adds convenience wrappers for l2 error and cosine similarity
to NS utils.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_extend_logger_results_with_comparison
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29600354](https://our.internmc.facebook.com/intern/diff/D29600354)

### CODE REVIEW

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29600354).

File: torch/quantization/ns/utils.py [hunk 1]
```diff
@@ -389,12 +389,46 @@ def maybe_add_missing_fqns(results: NSResultsType) -> None:
...
-        x = x.dequantize()
-    if y.is_quantized:
-        y = y.dequantize()
     Ps = torch.norm(x)
     Pn = torch.norm(x - y)
     return 20 * torch.log10(Ps / Pn)
+
+@maybe_dequantize_first_two_tensor_args_and_handle_tuples
+def compute_l2_error(x: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
+    return ((x - y) ** 2).sum()
```
Comment (hx89) [inline torch/quantization/ns/utils.py:425]: The l2 error we usually used is actually relative l2 error normalized by the l2 norm of the reference matrix x:

sqrt( ((x - y) ** 2).sum() / (x ** 2).sum() )

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29600354).

**Review by hx89 — APPROVED** [review]
lgtm!

**Comment by vkuzo** [comment]
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29600354).
